cpp/src/utils/metrics/yaml_config.cc

```diff
@@ -0,0 +1,192 @@
+#include "src/utils/metrics/yaml_config.hh"
+
+#include <set>
+
+#include "src/utils/logging.hh"
+
+namespace torchserve {
+void YAMLMetricsConfigurationHandler::LoadConfiguration(
+    const std::string metrics_config_file_path) {
+  ClearConfiguration();
+
+  try {
+    YAML::Node config_node = YAML::LoadFile(metrics_config_file_path);
+    ParseDimensionNames(config_node);
+    ParseSystemMetrics(config_node);
+    ParseModelMetrics(config_node);
+  } catch (YAML::ParserException& e) {
+    TS_LOGF(ERROR, "[METRICS]Failed to parse YAML configuration file: {}. {}",
+            metrics_config_file_path, e.what());
+    ClearConfiguration();
+  } catch (YAML::RepresentationException& e) {
+    TS_LOGF(ERROR,
+            "[METRICS]Failed to interpret YAML configuration file: {}. {}",
+            metrics_config_file_path, e.what());
+    ClearConfiguration();
+  } catch (YAML::Exception& e) {
+    TS_LOGF(ERROR, "[METRICS]Failed to load YAML configuration file: {}. {}",
+            metrics_config_file_path, e.what());
+    ClearConfiguration();
+  }
+}
+
+std::set<std::string> YAMLMetricsConfigurationHandler::GetDimensionNames() {
+  return dimension_names;
+}
+
+std::vector<MetricConfiguration>
+YAMLMetricsConfigurationHandler::GetSystemMetrics() {
+  return system_metrics;
+}
+
+std::vector<MetricConfiguration>
+YAMLMetricsConfigurationHandler::GetModelMetrics() {
+  return model_metrics;
+}
+
+void YAMLMetricsConfigurationHandler::ClearConfiguration() {
+  dimension_names.clear();
+  system_metrics.clear();
+  model_metrics.clear();
+}
+
+void YAMLMetricsConfigurationHandler::ParseDimensionNames(
+    const YAML::Node& document_node) {
+  if (!document_node["dimensions"]) {
+    return;
+  }
+
+  const std::vector<std::string> dimension_names_list =
+      document_node["dimensions"].as<std::vector<std::string>>();
+
+  for (const auto& name : dimension_names_list) {
+    if (dimension_names.insert(name).second == false) {
+      throw YAML::Exception(YAML::Mark::null_mark(),
+                            "Dimension names defined under central "
+                            "\"dimensions\" key must be unique");
+    }
+    if (name.empty()) {
+      throw YAML::Exception(
+          YAML::Mark::null_mark(),
+          "Dimension names defined under \"dimensions\" key cannot be emtpy");
+    }
+  }
+}
+
+void YAMLMetricsConfigurationHandler::ParseSystemMetrics(
```

Comment:
There is high code redundancy b/w ParseSystemMetrics and ParseModelMetrics. Please refactor these parts